presto-main/src/main/java/com/facebook/presto/operator/repartition/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,127 @@
+import static sun.misc.Unsafe.ARRAY_LONG_INDEX_SCALE;
+
+public class LongArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    @VisibleForTesting
+    static final int POSITION_SIZE = Long.BYTES + Byte.BYTES;
+
+    private static final String NAME = "LONG_ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(LongArrayBlockEncodingBuffers.class).instanceSize();
+
+    private byte[] valuesBuffer;
+    private int valuesBufferIndex;
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        valuesBufferIndex = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateSerializedRowSizes(int[] serializedRowSizes)
+    {
+        throw new UnsupportedOperationException("accumulateSerializedRowSizes is not supported for fixed width types");
+    }
+
+    @Override
+    protected void accumulateSerializedRowSizes(int[] positionOffsets, int positionCount, int[] serializedRowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            serializedRowSizes[i] += (positionOffsets[i + 1] - positionOffsets[i]) * POSITION_SIZE;
+        }
+    }
+
+    @Override
+    public void appendDataInBatch()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendValuesToBuffer();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    @Override
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        output.writeInt(bufferedPositionCount);
+
+        serializeNullsTo(output);
+
+        if (valuesBufferIndex > 0) {
+            output.appendBytes(valuesBuffer, 0, valuesBufferIndex);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPositionsRetainedSizeInBytes() +
+                sizeOf(valuesBuffer) +
+                getNullsBufferRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                getNullsBufferSerializedSizeInBytes() +   // nulls buffer
+                valuesBufferIndex;              // values buffer
+    }
+
+    private void appendValuesToBuffer()
+    {
+        valuesBuffer = ensureCapacity(valuesBuffer, valuesBufferIndex + batchSize * ARRAY_LONG_INDEX_SCALE, LARGE, true);
+
+        int[] positions = getPositions();
+        if (decodedBlock.mayHaveNull()) {
+            for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+                int position = positions[i];
+                int newIndex = setLongUnchecked(valuesBuffer, valuesBufferIndex, decodedBlock.getLong(position));
+
+                // Make sure the if block contains only one instruction so that it compiles to a conditional move (cmov)
```

Comment:
Sure. The statement doesn't have to be a mov instruction, it can also be arithmetic. So I rephrased it a bit:
`// Make sure the branch statement contains only one instruction, so that JVM can compile to a conditional move (cmov)`